File: packages/gbnf-experiment/data/20260909T005215Z_9a3e20fe/ported_implementation/src/grammarGraph/getInputAsCodePoints.ts
import type { ValidInput } from './grammarGraphTypes.ts';

export const getCodePoint = (char: string): number => {
  const codePoint = char.codePointAt(0);
  if (codePoint === undefined) {
    throw new Error(`Could not get code point for character: ${char}`);
  }
  return codePoint;
};

export const getInputAsCodePoints = (src: ValidInput): number[] => {
  if (typeof src === 'number') {
    return [src];
  }

  if (Array.isArray(src)) {
    for (const c of src) {
      if (!Number.isInteger(c)) {
        throw new Error(`code_point must be an integer for ${c} if src is a list`);
      }
    }
    return src;
  }

  if (typeof src === 'string') {
    // Iterating a string yields code points, matching Python's `for s in src`.
    return [...src].map(getCodePoint);
  }

  throw new Error(`Invalid input type: ${typeof src}`);
};

File: proposed-projects/round-trip-experiment/reverse/20260909T022216Z_19e0fe30/20260910T055411Z_259d4a57/ported_implementation/gbnf/grammar_graph/get_input_as_code_points.py
from typing import List

from .grammar_graph_types import ValidInput


def get_code_point(char: str) -> int:
    if not char:
        raise ValueError(f"Could not get code point for character: {char}")
    code_point = ord(char[0])
    if not isinstance(code_point, int):
        raise ValueError("code_point must be an integer!")
    return code_point


def get_input_as_code_points(src: ValidInput) -> List[int]:
    if isinstance(src, bool):
        raise ValueError(f"Invalid input type: {type(src)}")

    if isinstance(src, int):
        return [src]

    if isinstance(src, list):
        for c in src:
            if not isinstance(c, int) or isinstance(c, bool):
                raise ValueError(
                    f"code_point must be an integer for {c} if src is a list"
                )
        return src

    if isinstance(src, str):
        # iterating a string yields whole code points
        return [get_code_point(c) for c in src]

    raise ValueError(f"Invalid input type: {type(src)}")
